larray/inout/hdf.py

```diff
@@ -70,46 +70,54 @@ def read_hdf(filepath_or_buffer, key, fill_value=nan, na=nan, sort_rows=False, s
-            res = Axis(labels=labels, name=name)
-            res._iswildcard = attrs['wildcard']
-        elif _type == 'Group':
-            if name is None:
-                name = str(pd_obj.name)
-            if name == 'None':
-                name = None
-            key = pd_obj.values
-            if 'dtype_kind' in attrs and attrs['dtype_kind'] == 'U':
-                key = np.char.decode(key, 'utf-8')
-            axis = read_hdf(filepath_or_buffer, attrs['axis_key'])
-            res = LGroup(key=key, name=name, axis=axis)
+        try:
+            pd_obj = store.get(key)
+            attrs = store.get_storer(key).attrs
+            writer = attrs.writer if 'writer' in attrs else None
+            # for backward compatibility but any object read from an hdf file should have an attribute 'type'
+            _type = attrs.type if 'type' in attrs else 'Array'
+            _meta = attrs.metadata if 'metadata' in attrs else None
+            if _type == 'Array':
+                # cartesian product is not necessary if the array was written by LArray
+                cartesian_prod = writer != 'LArray'
+                res = df_asarray(pd_obj, sort_rows=sort_rows, sort_columns=sort_columns, fill_value=fill_value,
+                                 parse_header=False, cartesian_prod=cartesian_prod)
+                if _meta is not None:
+                    res.meta = _meta
+            elif _type == 'Axis':
+                if name is None:
+                    name = str(pd_obj.name)
+                if name == 'None':
+                    name = None
+                labels = pd_obj.values
+                if 'dtype_kind' in attrs and attrs['dtype_kind'] == 'U':
+                    # this check is there because there are cases where dtype_kind is 'U' but pandas returns
+                    # an array with object dtype containing bytes instead of a string array, and in that case
+                    # np.char.decode does not work
+                    # this is at least the case for Python2 + Pandas 0.24.2 combination
+                    if labels.dtype.kind == 'O':
+                        labels = np.array([l.decode('utf-8') for l in labels], dtype='U')
+                    else:
+                        labels = np.char.decode(labels, 'utf-8')
+                res = Axis(labels=labels, name=name)
+                res._iswildcard = attrs['wildcard']
+            elif _type == 'Group':
+                if name is None:
+                    name = str(pd_obj.name)
+                if name == 'None':
+                    name = None
+                key = pd_obj.values
+                if 'dtype_kind' in attrs and attrs['dtype_kind'] == 'U':
+                    key = np.char.decode(key, 'utf-8')
+                axis = read_hdf(filepath_or_buffer, attrs['axis_key'])
+                res = LGroup(key=key, name=name, axis=axis)
+        except KeyError:
+            scalar_key = key
+            attrs = store.root._v_attrs
```

Comment:
I would **MUCH** rather have scalars stored in a dedicated group (e.g. `__scalars__`) like for Axis and Group objects, than in the root node. AFAICT, this would already avoid conflicts for scalars named "metadata", "writer" or "type". Also, I would prefer (but I do **NOT** know if it is practical to do so!) if scalars were each stored as a standalone object (probably either a 0d array, or a 1d array of len 1) instead of being stored all as **attributes** of the root node. AFAICT, this would solve all other remaining name conflicts (with "system" attributes -- _v_attrnamessys) and would be conceptually cleaner: a scalar is some kind of data, not *meta*data, like attributes are.

Reply:
From the description of the "obj" argument of the table.Array class (at http://www.pytables.org/usersguide/libref/homogenous_storage.html#tables.Array), it seems like Python scalars might be natively supported by pytables...